_posts/2021-03-17-encoding-effects-with-freer-simple.markdown

```diff
@@ -0,0 +1,474 @@
+---
+layout: post
+title: "Encoding Effects using freer-simple"
+author: Juan Pablo Royo Sales
+date: 2021-03-17
+excerpt: Organize your program encoding effects using freer-simple library. 
+categories: [haskell, fp, functional, effects, effects system, effects encoding, effect, encoding, type level programming, types, functional programming]
+tags: [haskell, fp, functional, effects, effects system, effects encoding, effect, encoding, type level programming, types, functional programming]
+permalink: /posts/2021-03-17-encoding-effects-freer-simple.html
+---
+
+## Introduction
+
+In previous posts I have [written](/posts/2019-03-17-tagless-final-haskell.html) about how encode **Effects** using a **Tagless Final** approach. 
+On that article, the idea was to show how to organize a program which can have different *Effects* and compose and combine them in a single program. I've also explored the *technique* of **Tagless Final** for representing the encoding and [**mtl**](https://hackage.haskell.org/package/mtl) in particular to *handle or interpret* effects.
+
+## Context
+
+**Effectful computations** or **Algebraic Handlers** is a quite interesting topic for me, because in the majority of Industry Software Solutions nowadays it is impossible to escape from connecting with some external Service or System. Apart from that, in my opinion, any Industry Software Solutions should have at least *Metrics* and *Logs* before be shipped into production. Saying that, it is almost impossible to escape from this and we need to deal with side effects, in one way or another. 
+
+Everyone of us, as **Functional Programming Developer** want to deal only with *pure* functions and we don't want to deal with *side effects* or *non-deterministic* computations. But *pure FP* can happen only on small portions of our software solutions in the Industry and at the end we need to deal with *side effects*. 
```

Comment:
Technically nondeterminism is it a type of side effect.

Reply:
Effectful code is pure per se since it has no side effects but I get what you meant. Maybe add a footnote (?)